contracts/data-storage/andromeda-form/src/execute.rs

```diff
@@ -0,0 +1,418 @@
+            Some(saved_start_time) => {
+                if saved_start_time.gt(&end_time) {
+                    return Err(ContractError::CustomError {
+                        msg: format!("Not opened yet. Will be opend at {:?}", saved_start_time),
+                    });
+                } else if saved_end_time.gt(&end_time) {
+                    config.end_time = Some(end_time);
+                    CONFIG.save(ctx.deps.storage, &config)?;
+                } else {
+                    return Err(ContractError::CustomError {
+                        msg: format!("Already closed. Closed at {:?}", saved_end_time),
+                    });
+                }
+            }
+            None => {
+                return Err(ContractError::CustomError {
+                    msg: "Not opened yet".to_string(),
+                })
+            }
+        },
+        None => match saved_start_time {
+            Some(saved_start_time) => {
+                if saved_start_time.gt(&end_time) {
+                    return Err(ContractError::CustomError {
+                        msg: format!("Not opened yet. Will be opend at {:?}", saved_start_time),
+                    });
+                } else {
+                    config.end_time = Some(end_time);
+                    CONFIG.save(ctx.deps.storage, &config)?;
+                }
+            }
+            None => {
+                return Err(ContractError::CustomError {
+                    msg: "Not opened yet".to_string(),
+                })
+            }
+        },
+    }
+
+    let response = Response::new()
+        .add_attribute("method", "close_form")
+        .add_attribute("sender", sender);
+
+    Ok(response)
+}
+
+pub fn milliseconds_from_expiration(expiration: Expiration) -> Result<Milliseconds, ContractError> {
+    match expiration {
+        Expiration::AtTime(time) => Ok(Milliseconds::from_nanos(time.nanos())),
+        _ => Err(ContractError::CustomError {
+            msg: "Not supported expiration enum".to_string(),
+        }),
+    }
+}
+
+pub fn validate_form_is_opened(
+    current_time: Milliseconds,
+    saved_start_time: Option<Milliseconds>,
+    saved_end_time: Option<Milliseconds>,
+) -> Result<(), ContractError> {
+    match saved_start_time {
+        Some(saved_start_time) => {
+            if saved_start_time.gt(&current_time) {
+                return Err(ContractError::CustomError {
+                    msg: format!("Not opened yet. Will be opend at {:?}", saved_start_time),
+                });
+            }
+            if let Some(saved_end_time) = saved_end_time {
+                if current_time.gt(&saved_end_time) {
+                    return Err(ContractError::CustomError {
+                        msg: format!("Already closed. Closed at {:?}", saved_end_time),
+                    });
+                }
+            }
+            Ok(())
+        }
+        None => Err(ContractError::CustomError {
```

Comment:
Shouldn't we consider this that the user wants the form to be opened from the start?

Reply:
The start time will be set at instantiation or only owner and operator can open/close the form.